### PR TITLE
bpf: migrate hand-written bpf syscalls to ebpf-go APIs

### LINKED ISSUE #22513: Migrate pkg/bpf to use cilium/ebpf
`pkg/bpf` is showing its age, and is generally considered to be a bit of a tangle. It's really challenging to extend or modify if you're not familiar with the underlying kernel APIs. It exposes syscall wrappers for most map operations, and has extremely poor separation of concerns. (e.g. most map ops take Map.fd and pass it to a syscall wrapper)

This should be replaced by an easy-to-use abstraction that allows for better composition. For example, bpf.Map should probably be an interface so caches, metrics, etc. can be implemented as wrappers instead of bolting everything directly onto bpf.Map. This is TBD based on the code I encounter in the process.

In a few afternoons of hacking, I was able to remove the old map creation, pinning and pin-loading behaviours and replaced it with `cilium/ebpf` equivalents. I remain cautiously optimistic for the remainder of the work. The initial/previous approach was to create a new package `pkg/ebpf` that was based on cilium/ebpf, but this would give us the worst of both worlds: reimplementing all existing `pkg/bpf` behaviours (and there are quite a few) _and_ porting all callers to use the new `pkg/ebpf`, which will require significant changes anyway.

This issue is for taking the other approach and gradually replacing `pkg/bpf`'s primitives while retaining as much of its existing behaviour/features as possible.

### LINKED ISSUE #23558: Remove `TestDummyProg()`
As a follow up to #23557 we shouldremove `TestDummyProg()`. Like @ti-mo mentioned in https://github.com/cilium/cilium/issues/20739#issuecomment-1331855419_ feature probing prog/attachtype combinations looses value if the new implementation uses `bpf_link`.

Instead of probing for a succeeding `PROG_ATTACH` we should just ask for forgiveness with the new logic in `pkg/socketlb` and bail out or flip flags in cases of failure.

### PR DESCRIPTION
**Pull request by ti-mo**
This is based on https://github.com/cilium/cilium/pull/22693 and contains all of its commits. That one is close to being merged, so this can be rebased later.

See individual commits. Again a bit of churn, but relatively straightforward.

Fixes: https://github.com/cilium/cilium/issues/22513
Fixes: https://github.com/cilium/cilium/issues/23558

```release-note
Replace legacy bpf syscalls with ebpf-go library APIs
```

### CODE REVIEW

**Review by nbusseneau — APPROVED** [review]

File: pkg/bpf/map_linux.go [hunk 1]
```diff
@@ -941,34 +938,55 @@ func (m *Map) deleteMapEntry(key MapKey, ignoreMissing bool) (deleted bool, err
...
+	}
+
+	if err != nil {
+		return false, fmt.Errorf("unable to delete element %s from map %s: %w", key, m.name, err)
+	}
+
+	return true, nil
 }
 
 // SilentDelete deletes the map entry corresponding to the given key.
```
Comment (lmb) [inline pkg/bpf/map_linux.go:954]: Drive by: this description doesn't tell me at all what makes the function so silent.

File: pkg/bpf/map_linux.go [hunk 2]
```diff
@@ -941,34 +938,55 @@ func (m *Map) deleteMapEntry(key MapKey, ignoreMissing bool) (deleted bool, err
...
-// If a map entry is not found this returns (true, nil).
+// If a map entry is not found this returns (false, nil).
+// Takes out a lock on the Map.
```
Comment (lmb) [inline pkg/bpf/map_linux.go]: Why is this relevant to the caller? Can they somehow lock the map themselves?

File: pkg/maps/cidrmap/cidrmap.go [hunk 3]
```diff
@@ -42,7 +42,12 @@ const (
 
 type cidrKey struct {
 	Prefixlen uint32
-	Net       [16]byte
+
+	// v4 LPM maps have 8-byte key sizes even though the 20-byte cidrKey is used
+	// for map ops. This hack relies on passing unsafe.Pointers to the cidrKey so
```
Comment (lmb) [inline pkg/maps/cidrmap/cidrmap.go:47]: Oh boy

File: pkg/bpf/map_linux.go [hunk 4]
```diff
@@ -1023,42 +986,29 @@ func (m *Map) DeleteAll() error {
...
-	var err error
-	defer m.deleteAllMapEvent(err)
-	for {
-		if err := GetFirstKey(m.FD(), unsafe.Pointer(&nextKey[0])); err != nil {
-			if errors.Is(err, io.EOF) {
-				return nil
-			}
-			return err
-		}
+	//TODO(tb): This method ignores err, pass nil instead.
```
Comment (lmb) [inline pkg/bpf/map_linux.go]: ?

File: pkg/bpf/bpf.go [hunk 5]
```diff
@@ -19,11 +19,8 @@ var (
 const (
 	// BPF syscall command constants. Must match enum bpf_cmd from linux/bpf.h
 	// Deprecated: don't implement syscalls directly, use cilium/ebpf.
-
-	BPF_MAP_LOOKUP_ELEM = 1
```
Comment (lmb) [inline pkg/bpf/bpf.go]: Suuuper nit: this should've been removed in the previous commit?

File: pkg/datapath/linux/probes/attach_cgroup.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+	p, err := ebpf.NewProgramWithOptions(spec, ebpf.ProgramOptions{
+		LogDisabled: true,
+	})
+	if err != nil {
+		return fmt.Errorf("create cgroup program: %w: %w", err, ebpf.ErrNotSupported)
+	}
+	defer p.Close()
+
+	// Create arbitrary tempdir that's not a valid cgroup. Attaching to this sort
```
Comment (lmb) [inline pkg/datapath/linux/probes/attach_cgroup.go]: Out of curiosity: does it have to be a directory or is any file descriptor fine, e.g. `/dev/null`?

**Review by lmb — APPROVED** [review]

File: pkg/bpf/map_linux.go (hunk 1 above)
Comment (ti-mo) [reply pkg/bpf/map_linux.go:954]: Yeah, this needs to go. As you can see, the Map API is full of this sort of snowflake and unhelpful/incomplete doc. To answer your question: it skips incrementing the map's delete metric.

IMO this highlights a common problem with the design of the data model in the upper layer(s). The data (e.g. ct record) and the underlying data store (e.g. the ct map) are conflated. The bpf layer shouldn't lie about the amount of deletes it issues to the kernel, as it could mask a potential perf problem.

File: pkg/bpf/map_linux.go (hunk 4 above)
Comment (ti-mo) [reply pkg/bpf/map_linux.go]: Yep, just went ahead and removed the param, thanks.

File: pkg/bpf/bpf.go (hunk 5 above)
Comment (ti-mo) [reply pkg/bpf/bpf.go]: Nice catch! :grin:

File: pkg/datapath/linux/probes/attach_cgroup.go (hunk 6 above)
Comment (ti-mo) [reply pkg/datapath/linux/probes/attach_cgroup.go]: /dev/null works great, thanks!
